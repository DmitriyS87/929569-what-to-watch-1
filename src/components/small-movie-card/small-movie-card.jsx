import React from 'react';
import PropTypes from 'prop-types';


class SmallMovieCard extends React.PureComponent {

  constructor(props) {
    super(props);

    this._onPlay = this._onPlay.bind(this);
    this._onFocus = this._onFocus.bind(this);
    this._onBlur = this._onBlur.bind(this);
  }

  _onPlay() {
    this.props.onPlayClick(this);
  }

  _onFocus() {
    this.props.onFocus(this.props.movie.id);
  }

  _onBlur() {
    this.props.onBlur();
  }

  render() {
    const {movie} = this.props;

    return (
      <article key={`movie` + movie.id} className="small-movie-card catalog__movies-card" onFocus={this._onFocus} onBlur={this._onBlur} onMouseEnter={this._onFocus} onMouseLeave={this._onBlur} >
        <button className="small-movie-card__play-btn" type="button" onClick={this._onPlay}>Play</button>
        <div className="small-movie-card__image">
          <img src={movie.coverSrc} alt={movie.title} width="280" height="175" />
        </div>
        <h3 className="small-movie-card__title">
          <a className="small-movie-card__link" href="movie-page.html">{movie.title}</a>
        </h3>
      </article>
    );
  }
}

SmallMovieCard.propTypes = {
  movie: PropTypes.shape({
    title: PropTypes.string.isRequired,
    coverSrc: PropTypes.string,
    id: PropTypes.number.isRequired
  }),
  onPlayClick: PropTypes.func,
  onFocus: PropTypes.func.isRequired,
  onBlur: PropTypes.func.isRequired,
};

export default SmallMovieCard;
