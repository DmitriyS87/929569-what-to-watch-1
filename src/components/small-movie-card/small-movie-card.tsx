import * as React from 'react';
import VideoPlayer from '../video-player/video-player';
import { Movie } from '../../types';
import { withRouter } from 'react-router-dom';

interface Props {
  onFocus: (param: number) => void,
  movie: Movie,
  active: boolean,
  onBlur: () => void,
  history: { push: (param: string) => void },
}

class SmallMovieCard extends React.PureComponent<Props> {
  private timeOutId: any;
  constructor(props) {
    super(props);

    this._onFocus = this._onFocus.bind(this);
    this._onBlur = this._onBlur.bind(this);
    this._clearTimeOut = this._clearTimeOut.bind(this);
    this._handlerClick = this._handlerClick.bind(this);
  }

  _startTimer() {
    const self = this;
    const onPlay = this.props.onFocus;
    const startPlay = (movieCard: this, callback: (number) => void) => {
      callback(movieCard.props.movie.id);
    };
    this.timeOutId = setTimeout(startPlay, 1000, self, onPlay);
  }

  _onFocus() {
    this._startTimer();
  }

  _onBlur() {
    this.props.onBlur();
    this._clearTimeOut();
  }

  _clearTimeOut() {
    if (this.timeOutId) {
      clearTimeout(this.timeOutId);
    }
  }

  _handlerClick(evt) {
    const { history, movie } = this.props;
    evt.preventDefault();
    history.push(`/film/:${movie.id}`);
  }

  render() {
    const { movie, active } = this.props;

    const screen = () => {

      if (active) {
        return <VideoPlayer poster={movie.previewImg.src} movies={[movie.previewMovie]} title={movie.title} onMouseLeave={this._onBlur} />;
      }

      return (<React.Fragment>
        <div className="small-movie-card__image">
          <img src={movie.previewImg.src} alt={movie.title} width="280" height="175" />
        </div>
        <h3 className="small-movie-card__title">
          <a className="small-movie-card__link" href={`/film/:${movie.id}`}>{movie.title}</a>
        </h3>
      </React.Fragment>);
    };

    return (
      <article key={`movie` + movie.id} className="small-movie-card catalog__movies-card" onClick={this._handlerClick} onMouseEnter={this._onFocus} onMouseLeave={this._onBlur} >
        {screen()}
      </article>
    );
  }

  componentWillUnmount() {
    this._clearTimeOut();
  }
}

export default withRouter(SmallMovieCard);
export { SmallMovieCard };
