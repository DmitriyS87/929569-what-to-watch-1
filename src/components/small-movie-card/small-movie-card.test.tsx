import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { SmallMovieCard } from './small-movie-card';
import TEST_MOCKS from '../../mocks/test-mocks.js';

it(`SmallMovieCard correctly renders after relunch`, () => {
  const mockHandler = jest.fn();
  const MOCK_ACTIVE = false;
  const tree = renderer
    .create(<SmallMovieCard
      movie={TEST_MOCKS.film}
      active={MOCK_ACTIVE}
      onFocus={mockHandler}
      onBlur={mockHandler}
      history={{ push: jest.fn() }}
    />)
    .toJSON();

  expect(tree).toMatchSnapshot();
});
