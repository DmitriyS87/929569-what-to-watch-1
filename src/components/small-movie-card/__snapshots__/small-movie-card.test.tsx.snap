// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SmallMovieCard correctly renders after relunch 1`] = `
<article
  className="small-movie-card catalog__movies-card"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <div
    className="small-movie-card__image"
  >
    <img
      alt="Aviator"
      height="175"
      src="img/aviator.jpg"
      width="280"
    />
  </div>
  <h3
    className="small-movie-card__title"
  >
    <a
      className="small-movie-card__link"
      href="/film/:0"
    >
      Aviator
    </a>
  </h3>
</article>
`;
